pkg/controllers/consolidation/controller.go

```diff
@@ -0,0 +1,665 @@
+	// further consolidation.
+	var once sync.Once
+	if err := retry.Do(func() error {
+		if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: nodeNames[0]}, &k8Node); err != nil {
+			return fmt.Errorf("getting node, %w", err)
+		}
+		once.Do(func() {
+			c.recorder.LaunchingNodeForConsolidation(&k8Node, minCost.String())
+		})
+
+		if _, ok := k8Node.Labels[v1alpha5.LabelNodeInitialized]; !ok {
+			// make the user aware of why consolidation is paused
+			c.recorder.WaitingOnReadinessForConsolidation(&k8Node)
+			return errors.New("node is not initialized")
+		}
+		return nil
+	}, retry.Delay(2*time.Second),
+		retry.LastErrorOnly(true),
+		retry.Attempts(30),
+		retry.MaxDelay(10*time.Second), // ~ 4.5 minutes in total
+	); err != nil {
+		// node never become ready, so uncordon the node we were trying to delete and report the error
+		return multierr.Combine(c.setNodeUnschedulable(ctx, minCost.oldNodes[0].Name, false),
+			fmt.Errorf("checking node readiness, %w", err))
+	}
+	return nil
+}
+
+func (c *Controller) getNodePods(ctx context.Context, nodeName string) ([]*v1.Pod, error) {
+	var podList v1.PodList
+	if err := c.kubeClient.List(ctx, &podList, client.MatchingFields{"spec.nodeName": nodeName}); err != nil {
+		return nil, fmt.Errorf("listing pods, %w", err)
+	}
+	var pods []*v1.Pod
+	for i := range podList.Items {
+		// these pods don't need to be rescheduled
+		if pod.IsOwnedByNode(&podList.Items[i]) ||
+			pod.IsOwnedByDaemonSet(&podList.Items[i]) ||
+			pod.IsTerminal(&podList.Items[i]) {
+			continue
+		}
+		pods = append(pods, &podList.Items[i])
+	}
+	return pods, nil
+}
+
+func (c *Controller) canBeTerminated(node candidateNode, pdbs *PDBLimits) error {
+	if !node.DeletionTimestamp.IsZero() {
+		return fmt.Errorf("already being deleted")
+	}
+	if !pdbs.CanEvictPods(node.pods) {
+		return fmt.Errorf("not eligible for termination due to PDBs")
+	}
+	return c.podsPreventEviction(node)
+}
+
+func (c *Controller) podsPreventEviction(node candidateNode) error {
+	for _, p := range node.pods {
+		// don't care about pods that are finishing, finished or owned by the node
+		if pod.IsTerminating(p) || pod.IsTerminal(p) || pod.IsOwnedByNode(p) {
+			continue
+		}
+
+		if pod.HasDoNotEvict(p) {
+			return fmt.Errorf("found do-not-evict pod")
+		}
+
+		if pod.IsNotOwned(p) {
+			return fmt.Errorf("found pod with no controller")
+		}
+	}
+	return nil
+}
+
+type consolidateResult byte
+
+const (
+	consolidateResultUnknown = iota
+	consolidateResultNotPossible
+	consolidateResultDelete
+	consolidateResultDeleteEmpty
+	consolidateResultShrink
+)
+
+func (r consolidateResult) String() string {
+	switch r {
+	case consolidateResultUnknown:
+		return "Unknown"
+	case consolidateResultNotPossible:
+		return "Not Possible"
+	case consolidateResultDelete:
+		return "Delete"
+	case consolidateResultDeleteEmpty:
+		return "Delete (empty node)"
+	case consolidateResultShrink:
+		return "Shrink"
```

Comment:
Since this is a customer facing term, I'd like to advocate for calling this "Replace". Specifically, we won't necessarily be shrinking the node, especially not across all dimensions. i.e., I'd expect something in our  docs like "Karpenter will replace a node if the workloads on it would be supported by a cheaper one."

Reply:
Seems fine to me.